docker_publisher.py

```diff
@@ -0,0 +1,299 @@
+        of the image for the product."""
+        raise "pure virtual"
+
+    def getDockerImage(self, callback):
+        """This function downloads the root fs layer and calls callback
+        with its path as argument."""
+        raise "pure virtual"
+
+
+class DockerFetchException(Exception):
+    pass
+
+
+class DockerImagePublisherGit(DockerImagePublisher):
+    def __init__(self, git_path, git_branch, path, arch):
+        """Initialize a DockerImagePublisherGit with:
+        @git_path: Path to the local git repo clone
+        @git_branch: Branch of the image
+        @path: Path to the directory within the branch which will contain:
+        - Dockerfile (including version in a comment)
+        - *.tar.xz: Images"""
+        self.git_path = git_path
+        self.git_branch = git_branch
+        self.path = path
+        self.arch = arch
+
+        self.git_call = ["git", "-C", self.git_path]
+
+        ret = subprocess.call(self.git_call + ["fetch", "origin"])
+        if ret != 0:
+            raise DockerFetchException("Could not fetch from origin")
+
+    def releasedDockerImageVersion(self):
+        # Read from git using cat-file to avoid expensive checkout.
+        args = self.git_call + ["cat-file", "--textconv", "origin/%s:%s/Dockerfile" % (self.git_branch, self.path)]
+        with subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as p:
+            version_regex = re.compile("^# Version: (.+)$")
+            for line in p.stdout:
+                match = version_regex.match(line.decode('utf-8').strip())
+                if match:
+                    return match.group(1)
+
+        return "0"
+
+    def generateDockerFile(self, version, filename):
+        """Return the contents of a docker file for the image"""
+        return """FROM scratch
+MAINTAINER Fabian Vogt <fvogt@suse.com>
+# Version: %s
+ADD %s /
+""" % (version, filename)
+
+    def releaseDockerImage(self, version, rootfile):
+        # Try to delete the old branch
+        try:
+            subprocess.call(self.git_call + ["checkout", "-q", "origin/%s" % (self.git_branch)])
+            subprocess.check_output(self.git_call + ["branch", "-q", "-D", self.git_branch],
+                                    stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError:
+            pass
+
+        # We do an orphan checkout to not carry the history around
+        ret = subprocess.call(self.git_call + ["checkout", "-q", "--orphan", self.git_branch])
+        if ret != 0:
+            raise DockerPublishException("Could not checkout git branch")
+
+        # Remove all old files
+        for file in glob.glob("%s/%s/*" % (self.git_path, self.path)):
+            os.remove(file)
+
+        # Re-compress it into the correct location and format
+        targetfilename = "openSUSE-Tumbleweed-%s-%s.tar.xz" % (self.arch, version)
+
+        if not recompress(rootfile, "%s/%s/%s" % (self.git_path, self.path, targetfilename)):
```

Comment:
For making paths us `os.path.join()`. Up to you, but I like `from os import path` and `path.join()`.

Reply:
Which is actually longer and more cumbersome to use, this is not going to work on windows anyway...

What about `pathlib`?
It allows syntax like `self.git_path / self.path / targetfilename`, which seems like a good compromise.

Reply:
Does not appear to be in SLE SP3 repo which is where most of this code ends up running. Up to you, I never find:

```python
"%s/%s/%s" % (...)
```

more readable or shorter than

```python
path.join(...)
```

even

```python
'/'.join([...])
```

Reply:
Still not seeing the allure in manually writing out `%s/` for each part over either other solution, but the variable is nice.